autopush-common/src/db/bigtable/bigtable_client/mod.rs

```diff
@@ -505,9 +546,53 @@ impl DbClient for BigTableClientImpl {
     /// the individual cells create a new version. Depending on the garbage collection rules for
     /// the family, these can either persist or be automatically deleted.
     async fn update_user(&self, user: &User) -> DbResult<bool> {
-        self.add_user(user).await?;
-        // TODO: is a conditional check possible?
-        Ok(true)
+        // `update_user` supposes the following logic:
+        //   check to see if the user is in the router table AND
+        //   (either does not have a router_type or the router type is the same as what they're currently using) AND
+        //   (either there's no node_id assigned or the `connected_at` is earlier than the current connected_at)
+        // Bigtable can't really do a few of those (e.g. detect non-existing values) so we have to simplify
+        // things down a bit. We presume the record exists and that the `router_type` was specified when the
+        // record was created.
+        //
+        // Filters are not very sophisticated on BigTable.
+        // You can create RowFilterChains, where each filter acts as an "AND" or
+        // a RowFilterUnion which acts as an "OR".
+        //
+        // There do not appear to be negative checks (e.g. check if not set)
+        // According to [the docs](https://cloud.google.com/bigtable/docs/using-filters#chain)
+        // ConditionalRowFilter is not atomic and changes can occur between predicate
+        // and execution.
+        //
+        // We then use a ChainFilter (essentially an AND operation) to make sure that the router_type
+        // matches and the new `connected_at` time is later than the existing `connected_at`
+
+        let row = self.user_to_row(user);
+        let mut filter_chain = RowFilter_Chain::default();
+        let mut filter_set: RepeatedField<RowFilter> = RepeatedField::default();
+
+        // First check to see if the router type is either empty or matches exactly.
+        let mut type_filter = RowFilter::default();
+        type_filter.set_family_name_regex_filter(ROUTER_FAMILY.to_owned());
+        type_filter.set_column_qualifier_regex_filter("router_type".to_owned().as_bytes().to_vec());
+        type_filter.set_value_regex_filter(user.router_type.as_bytes().to_vec());
+        filter_set.push(type_filter);
+
+        // then check to make sure that the last connected_at time is before this one.
+        let mut connected_filter = RowFilter::default();
+        connected_filter.set_family_name_regex_filter(ROUTER_FAMILY.to_owned());
+        connected_filter
+            .set_column_qualifier_regex_filter("connected_at".to_owned().as_bytes().to_vec());
+        let mut val_range = ValueRange::default();
+        val_range.set_start_value_closed(user.connected_at.to_be_bytes().to_vec());
```

Comment:
_the new `connected_at` time is later than the existing `connected_at`_

doesn't that mean a:

```suggestion
        val_range.set_end_value_closed(user.connected_at.to_be_bytes().to_vec());
```
A test for this would be great (I'm not sure anything like the integration test actually test this) but I'm ok adding one after this PR if it's too involved.

Reply:
Sigh. I remember this. It's completely counter intuitive, and I'll add a comment to address it.
What the filter is doing is ensuring that the value that we will provide is ranged to be inclusive of the provided timestamp, not that the existing value is below that timestamp. Honestly, I shouldn't use this filter. 

I'm not sure if there's a way to specify "Only modify if this condition is present".

Reply:
Turns out, there is. Thanks @pjenvey.